python/gps/utility/generate_demo.py

```diff
@@ -0,0 +1,200 @@
+            """
+            # Load the algorithm
+
+            self.algorithms = self.load_algorithms()
+            self.algorithm = self.algorithms[0]
+
+            # Keep the initial states of the agent the sames as the demonstrations.
+            agent_config = self._hyperparams['demo_agent']
+            self.agent = agent_config['type'](agent_config)
+
+            # Roll out the demonstrations from controllers
+            var_mult = self._hyperparams['algorithm']['demo_var_mult']
+            T = self.algorithms[0].T
+            demos = []
+            demo_idx_conditions = []  # Stores conditions for each demo
+
+            M = agent_config['conditions']
+            N = self._hyperparams['algorithm']['num_demos']
+            if not self.nn_demo:
+                controllers = {}
+
+                # Store each controller under M conditions into controllers.
+                for i in xrange(M):
+                    controllers[i] = self.algorithm.cur[i].traj_distr
+                controllers_var = copy.copy(controllers)
+                for i in xrange(M):
+                    # Increase controller variance.
+                    controllers_var[i].chol_pol_covar *= var_mult
+                    # Gather demos.
+                    for j in xrange(N):
+                        demo = self.agent.sample(
+                            controllers_var[i], i,
+                            verbose=(j < self._hyperparams['verbose_trials']), noisy=True,
+                            save = True
+                        )
+                        demos.append(demo)
+                        demo_idx_conditions.append(i)
+            else:
+                all_pos_body_offsets = []
+                # Gather demos.
+                for a in xrange(len(self.algorithms)):
+                    pol = self.algorithms[a].policy_opt.policy
+                    for i in xrange(M / len(self.algorithms) * a, M / len(self.algorithms) * (a + 1)):
+                        for j in xrange(N):
+                            demo = self.agent.sample(
+                                pol, i,
+                                verbose=(j < self._hyperparams['verbose_trials']), noisy=True
+                                )
+                            demos.append(demo)
+                            demo_idx_conditions.append(i)
+
+            # Filter failed demos
+            if 'filter_demos' in  agent_config:
+                filter_options = agent_config['filter_demos']
+                filter_type = filter_options.get('type', 'min')
+                targets = filter_options['target']
+                pos_idx = filter_options['state_idx']
+                max_per_condition = filter_options.get('max_demos_per_condition', 999)
+                dist_threshold = filter_options.get('success_upper_bound', 0.01)
+                cur_samples = SampleList(demos)
+                sample_end_effectors = [cur_samples[i].get_X()[:,pos_idx] for i in xrange(len(cur_samples))]
```

Comment:
can you make a general util function for computing these distances? (and use it both here and in the algorithm files, and whereever else we compute this)